src/persist-client/examples/maelstrom/services.rs

```diff
@@ -0,0 +1,337 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Implementations of Maelstrom services as persist external durability
+
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::Instant;
+
+use anyhow::anyhow;
+use async_trait::async_trait;
+use mz_persist::location::{Atomicity, BlobMulti, Consensus, ExternalError, SeqNo, VersionedData};
+
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use tokio::sync::Mutex;
+
+use crate::maelstrom::api::{ErrorCode, MaelstromError};
+use crate::maelstrom::node::Handle;
+
+/// An adaptor for [VersionedData] that implements [Serialize] and
+/// [Deserialize].
+#[derive(Clone, Debug, Serialize, Deserialize)]
+struct MaelstromVersionedData {
+    seqno: u64,
+    data: Vec<u8>,
+}
+
+/// Implementation of [Consensus] backed by the Maelstrom lin-kv service.
+#[derive(Debug)]
+pub struct MaelstromConsensus {
+    handle: Handle,
+    // A cache of SeqNo -> the data for that SeqNo. Here because the [Consensus]
+    // interface uses only a SeqNo for expected/from, but the lin-kv Maelstrom
+    // service requires the whole VersionedData. This is only used to hydrate
+    // the expected/from in the impl of `compare_and_set`.
+    cache: Mutex<HashMap<(String, SeqNo), Vec<u8>>>,
+}
+
+impl MaelstromConsensus {
+    pub fn new(handle: Handle) -> Arc<dyn Consensus + Send + Sync> {
+        Arc::new(MaelstromConsensus {
+            handle,
+            cache: Mutex::new(HashMap::new()),
+        }) as Arc<dyn Consensus + Send + Sync>
+    }
+
+    pub async fn hydrate_seqno(
+        &self,
+        deadline: Instant,
+        key: &str,
+        expected: SeqNo,
+    ) -> Result<Result<VersionedData, Option<VersionedData>>, ExternalError> {
+        if let Some(data) = self.cache.lock().await.get(&(key.to_string(), expected)) {
+            let value = VersionedData {
+                seqno: expected.clone(),
+                data: data.clone(),
+            };
+            return Ok(Ok(value));
+        }
+
+        // It wasn't in the cache (must have been set by another process), fetch
```

Comment:
if i understand this code right (big if) i believe you could skip the cache and just always call `head` to see if the seqno matches and if so, use the appropriate expected value. 

I can't quite tell whether the cache is in place in order to:

1. exercise more invalid states where we compare_and_set with stale seqnos
2. be a perf optimization to reduce calls to head
3. some other reason